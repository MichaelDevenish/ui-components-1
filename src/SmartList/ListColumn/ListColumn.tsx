import React from 'react'
import { isNil, merge } from 'lodash'
import classNames from 'classnames'
import { Col } from 'react-styled-flexboxgrid'
import { ListHeader } from '../ListHeader'

export interface ISize {
  xs?: number,
  sm?: number,
  md?: number,
  lg?: number
}

export interface IListColumn {
  /** Content to display in each cell of the list column */
  cell: (rowObject: any) => JSX.Element | string
  /** Class names to use for each cell of the list column */
  cellClassNames?: (rowObject: any) => string
  /** Dataset used to generate the list content */
  data?: any
  /** Index of the row being iterated on */
  rowIndex?: number
  /** Flag to state if the column cell is the header or not */
  isHeader?: boolean
  /** Content to display in header */
  header?: string | object
  /** Size of the column */
  size?: ISize
  /** Flag to select if the text should be aligned right */
  alignRight?: boolean
  /** Text display in the column header tooltip */
  tooltipText?: string
  /** Foundation behave attribute */
  behave?: string
  /** Foundation order attribute */
  order?: any
}

export class ListColumn extends React.PureComponent<IListColumn> {
  public static defaultProps: Partial<IListColumn> = {
    rowIndex: 0,
    isHeader: false,
    header: '',
    alignRight: false
  }

  get cellContent (): JSX.Element | string | undefined {
    const {
      cell,
      data,
      rowIndex,
      isHeader,
      header,
      tooltipText
    } = this.props

    if (isHeader) {
      return (
        <ListHeader
          label={header}
          tooltipText={tooltipText}
        />
      )
    }

    if (!isNil(rowIndex) && data && data[rowIndex]) {
      return cell(data[rowIndex])
    }
  }

  get cellClassNames (): string | undefined {
    const {
      cellClassNames,
      data,
      rowIndex
    } = this.props

    if (!isNil(rowIndex) && data && data[rowIndex] && cellClassNames) {
      return cellClassNames(data[rowIndex])
    }
  }

  get sizeParams (): ISize {
    const {
      size
    } = this.props

    return merge({
      xs: 12,
      sm: 12,
      md: 12,
      lg: 12
    }, size)
  }

  public render (): JSX.Element {
    const {
      alignRight,
      size
    } = this.props

    const content = (
      <div className={classNames(
        'list-column',
        this.cellClassNames,
        {
          'text-right': alignRight
        }
      )}>
        {this.cellContent}
      </div>
    )

    if (size) {
      return (
        <Col {...this.sizeParams}>
          {content}
        </Col>
      )
    }

    return content
  }
}
