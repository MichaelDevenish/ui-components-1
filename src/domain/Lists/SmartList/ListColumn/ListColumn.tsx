import React from 'react'
import { isNil } from 'lodash'
import classNames from 'classnames'
import { Row } from '../../../Grids/Row'
import { ListHeader } from '../ListHeader'
import { TextSkeleton } from '../../../Skeletons'
import { ISmartListSkeletonOptions } from '../SmartList'

export interface ISize {
  xs?: number,
  sm?: number,
  md?: number,
  lg?: number
}

export interface IListColumn {
  /** Content to display in each cell of the list column */
  cell: (rowObject: any) => JSX.Element | string | number
  /** Class names to use for each cell of the list column */
  cellClassNames?: (rowObject: any) => string
  /** Dataset used to generate the list content */
  data?: any
  /** Index of the row being iterated on */
  rowIndex?: number
  /** Flag to state if the column cell is the header or not */
  isHeader?: boolean
  /** Content to display in header */
  header?: string | object
  /** Size of the column */
  size?: ISize
  /** Flag to select if the text should be aligned right */
  alignRight?: boolean
  /** Text display in the column header tooltip */
  tooltipText?: string
  /** Foundation behave attribute */
  behave?: string
  /** Foundation order attribute */
  order?: any
  /** Skeleton options */
  skeletonOptions: ISmartListSkeletonOptions
}

class ListColumn extends React.PureComponent<IListColumn> {
  public static defaultProps: Partial<IListColumn> = {
    rowIndex: 0,
    isHeader: false,
    header: '',
    alignRight: false,
    skeletonOptions: {
      showSkeleton: false
    }
  }

  get cellContent (): JSX.Element | string | undefined | number {
    const {
      cell,
      data,
      rowIndex,
      isHeader,
      header,
      tooltipText
    } = this.props

    const {
      showSkeleton
    } = this.props.skeletonOptions

    if (isHeader) {
      return (
        <ListHeader
          label={header}
          tooltipText={tooltipText}
        />
      )
    }

    if (showSkeleton) {
      return <TextSkeleton showSkeleton={showSkeleton} />
    }

    if (!isNil(rowIndex) && data && data[rowIndex]) {
      return cell(data[rowIndex])
    }
  }

  get cellClassNames (): string | undefined {
    const {
      cellClassNames,
      data,
      rowIndex
    } = this.props

    if (!isNil(rowIndex) && data && data[rowIndex] && cellClassNames) {
      return cellClassNames(data[rowIndex])
    }
  }

  public render (): JSX.Element {
    const {
      alignRight,
      size
    } = this.props

    const content = (
      <div className={classNames(
        'list-column',
        this.cellClassNames,
        {
          'text-right': alignRight
        }
      )}>
        {this.cellContent}
      </div>
    )

    if (size) {
      return (
        <Row.Column {...size}>
          {content}
        </Row.Column>
      )
    }

    return content
  }
}

export {
  ListColumn
}
