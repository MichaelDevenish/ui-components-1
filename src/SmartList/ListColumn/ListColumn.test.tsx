import React from 'react'
import { mount } from 'enzyme'
import { ListColumn } from './ListColumn'

export interface ITestListColumnItem {
  name: string
}

const testData = [
  {
    name: 'Test Name'
  }
]

describe('<ListColumn />', () => {
  it('should render a list column with the right non-header cell content', () => {
    const wrapper = mount(
      <ListColumn
        data={testData}
        rowIndex={0}
        size={6}
        cell={(row: ITestListColumnItem) =>
          <div>
            {row.name}
          </div>
        }
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render a list column with the right header', () => {
    const wrapper = mount(
      <ListColumn
        data={testData}
        rowIndex={0}
        size={6}
        isHeader
        header='Test'
        cell={(row: ITestListColumnItem) =>
          <div>
            {row.name}
          </div>
        }
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render a list column with custom column orders', () => {
    let wrapper = mount(
      <ListColumn
        data={testData}
        rowIndex={0}
        size={6}
        isHeader
        header='Test'
        order='large-order-2'
        cell={(row: ITestListColumnItem) =>
          <div>
            {row.name}
          </div>
        }
      />
    )

    expect(wrapper).toMatchSnapshot()

    wrapper = mount(
      <ListColumn
        data={testData}
        rowIndex={0}
        size={6}
        isHeader
        header='Test'
        order={2}
        cell={(row: ITestListColumnItem) =>
          <div>
            {row.name}
          </div>
        }
      />
    )

    expect(wrapper).toMatchSnapshot()

    wrapper = mount(
      <ListColumn
        data={testData}
        rowIndex={0}
        size={6}
        isHeader
        header='Test'
        order={{
          large: 2,
          small: 6
        }}
        cell={(row: ITestListColumnItem) =>
          <div>
            {row.name}
          </div>
        }
      />
    )

    expect(wrapper).toMatchSnapshot()
  })
})
