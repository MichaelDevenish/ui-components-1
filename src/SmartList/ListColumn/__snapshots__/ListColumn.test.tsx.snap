// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListColumn /> should render a list column with custom column orders 1`] = `
<ListColumn
  alignRight={false}
  cell={[Function]}
  data={
    Array [
      Object {
        "name": "Test Name",
      },
    ]
  }
  header="Test"
  isHeader={true}
  order="large-order-2"
  rowIndex={0}
  size={6}
>
  <div
    className="list-column cell small-6 large-order-2"
  >
    <ListHeader
      label="Test"
    >
      <span
        className="ListHeader"
      >
        Test
      </span>
    </ListHeader>
  </div>
</ListColumn>
`;

exports[`<ListColumn /> should render a list column with custom column orders 2`] = `
<ListColumn
  alignRight={false}
  cell={[Function]}
  data={
    Array [
      Object {
        "name": "Test Name",
      },
    ]
  }
  header="Test"
  isHeader={true}
  order={2}
  rowIndex={0}
  size={6}
>
  <div
    className="list-column cell small-6 small-order-2"
  >
    <ListHeader
      label="Test"
    >
      <span
        className="ListHeader"
      >
        Test
      </span>
    </ListHeader>
  </div>
</ListColumn>
`;

exports[`<ListColumn /> should render a list column with custom column orders 3`] = `
<ListColumn
  alignRight={false}
  cell={[Function]}
  data={
    Array [
      Object {
        "name": "Test Name",
      },
    ]
  }
  header="Test"
  isHeader={true}
  order={
    Object {
      "large": 2,
      "small": 6,
    }
  }
  rowIndex={0}
  size={6}
>
  <div
    className="list-column cell small-6 large-order-2 small-order-6"
  >
    <ListHeader
      label="Test"
    >
      <span
        className="ListHeader"
      >
        Test
      </span>
    </ListHeader>
  </div>
</ListColumn>
`;

exports[`<ListColumn /> should render a list column with the right header 1`] = `
<ListColumn
  alignRight={false}
  cell={[Function]}
  data={
    Array [
      Object {
        "name": "Test Name",
      },
    ]
  }
  header="Test"
  isHeader={true}
  rowIndex={0}
  size={6}
>
  <div
    className="list-column cell small-6"
  >
    <ListHeader
      label="Test"
    >
      <span
        className="ListHeader"
      >
        Test
      </span>
    </ListHeader>
  </div>
</ListColumn>
`;

exports[`<ListColumn /> should render a list column with the right non-header cell content 1`] = `
<ListColumn
  alignRight={false}
  cell={[Function]}
  data={
    Array [
      Object {
        "name": "Test Name",
      },
    ]
  }
  header=""
  isHeader={false}
  rowIndex={0}
  size={6}
>
  <div
    className="list-column cell small-6"
  >
    <div>
      Test Name
    </div>
  </div>
</ListColumn>
`;
